### PR TITLE
Make CompositeExpression immutable

### PR DESCRIPTION
**Pull request by BenMorel**
Continuing the work in #3850.

### CODE REVIEW

**Comment by BenMorel** [comment]
@morozov BTW, should this be targeting 3.0 or 4.0? I can see all these breaking changes in the UPGRADE file targeting 3.0, but IIRC you said that there would not be large BC breaks in 3.0? Did I understand this right?

**Comment by morozov** [comment]
These changes should be targeting master (4.0). The upgrade notes in `master` should be corrected have two different sections for `3.0` and `4.0`.

**Comment by BenMorel** [comment]
Do you mean that some of the changes marked as 3.0 in the UPGRADE file should target 4.0? Should you update this file first?

File: lib/Doctrine/DBAL/Query/Expression/CompositeExpression.php [hunk 1]
```diff
@@ -43,51 +47,10 @@ class CompositeExpression implements Countable
      */
     public function __construct(string $type, array $parts = [])
```
Comment (morozov) [inline lib/Doctrine/DBAL/Query/Expression/CompositeExpression.php]: In a future PR, I'd like the constructor to be made `private` in favor of two named constructors — `and()` and `or()` — which would accept `($part, ...$parts)` and therefore not allow construction of an empty composite expression.

Comment (BenMorel) [reply lib/Doctrine/DBAL/Query/Expression/CompositeExpression.php]: Yes, I've mentioned that this is temporary in https://github.com/doctrine/dbal/pull/3850#discussion_r370997009. I'd like to complete immutability first, then move on to stop accepting empty parts. And you're right, factory methods will be a good idea!

**Comment by morozov** [comment]
> Do you mean that some of the changes marked as 3.0 in the UPGRADE file should target 4.0? 

Yes.

> Should you update this file first?

https://github.com/doctrine/dbal/pull/3859

**Review by morozov — APPROVED** [review]

**Comment by morozov** [comment]
Thanks, @BenMorel!

**Pull request merged**